File: app/src/main/java/eu/kanade/mangafeed/ui/setting/preference/SourceLoginDialog.java
package eu.kanade.mangafeed.ui.setting.preference;

import android.content.Context;
import android.content.DialogInterface;
import android.view.View;

import eu.kanade.mangafeed.R;
import eu.kanade.mangafeed.data.preference.PreferencesHelper;
import eu.kanade.mangafeed.data.source.base.Source;
import eu.kanade.mangafeed.util.ToastUtil;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class SourceLoginDialog extends LoginDialogPreference {

    private Source source;

    public SourceLoginDialog(Context context, PreferencesHelper preferences, Source source) {
        super(context, preferences);
        this.source = source;
    }

    @Override
    protected void onBindDialogView(View view) {
        super.onBindDialogView(view);

        title.setText(getContext().getString(R.string.accounts_login_title, source.getName()));

        username.setText(preferences.getSourceUsername(source));
        password.setText(preferences.getSourcePassword(source));
    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        super.onDialogClosed(positiveResult);

        if (positiveResult) {
            preferences.setSourceCredentials(source,
                    username.getText().toString(),
                    password.getText().toString());
        }
    }

    protected void checkLogin() {
        if (requestSubscription != null)
            requestSubscription.unsubscribe();

        if (username.getText().length() == 0 || password.getText().length() == 0)
            return;

        loginBtn.setProgress(1);

        requestSubscription = source
                .login(username.getText().toString(), password.getText().toString())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(logged -> {
                    if (logged) {
                        // Simulate a positive button click and dismiss the dialog
                        onClick(dialog, DialogInterface.BUTTON_POSITIVE);
                        dialog.dismiss();
                        ToastUtil.showShort(context, R.string.login_success);
                    } else {
                        preferences.setSourceCredentials(source, "", "");
                        loginBtn.setProgress(-1);
                    }
                }, error -> {
                    loginBtn.setProgress(-1);
                    loginBtn.setText(R.string.unknown_error);
                });

    }

}

File: app/src/main/java/eu/kanade/tachiyomi/widget/preference/SourceLoginDialog.kt
package eu.kanade.tachiyomi.widget.preference

import android.os.Bundle
import android.view.View
import eu.kanade.tachiyomi.R
import eu.kanade.tachiyomi.source.Source
import eu.kanade.tachiyomi.source.SourceManager
import eu.kanade.tachiyomi.source.online.LoginSource
import eu.kanade.tachiyomi.util.toast
import kotlinx.android.synthetic.main.pref_account_login.view.*
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import uy.kohesive.injekt.Injekt
import uy.kohesive.injekt.api.get

class SourceLoginDialog(bundle: Bundle? = null) : LoginDialogPreference(bundle) {

    private val source = Injekt.get<SourceManager>().get(args.getLong("key")) as LoginSource

    constructor(source: Source) : this(Bundle().apply { putLong("key", source.id) })

    override fun setCredentialsOnView(view: View) = with(view) {
        dialog_title.text = context.getString(R.string.login_title, source.toString())
        username.setText(preferences.sourceUsername(source))
        password.setText(preferences.sourcePassword(source))
    }

    override fun checkLogin() {
        requestSubscription?.unsubscribe()

        v?.apply {
            if (username.text.isEmpty() || password.text.isEmpty())
                return

            login.progress = 1

            requestSubscription = source.login(username.text.toString(), password.text.toString())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ logged ->
                        if (logged) {
                            preferences.setSourceCredentials(source,
                                    username.text.toString(),
                                    password.text.toString())

                            dialog?.dismiss()
                            context.toast(R.string.login_success)
                        } else {
                            preferences.setSourceCredentials(source, "", "")
                            login.progress = -1
                        }
                    }, { error ->
                        login.progress = -1
                        login.setText(R.string.unknown_error)
                        error.message?.let { context.toast(it) }
                    })
        }
    }

    override fun onDialogClosed() {
        super.onDialogClosed()
        (targetController as? Listener)?.loginDialogClosed(source)
    }

    interface Listener {
        fun loginDialogClosed(source: LoginSource)
    }

}
